### ISSUE TITLE
Mocking a method of the classundertest

### ISSUE THREAD

**Issue by jmichas** [markjsc/AutoMockHelper#4]
How can I mock a method on the class I'm testing?
I need to mock the response of a service call that takes place in a method other than the one I'm testing.

More detail:
My LoginService has a method LoginUserAsync(string username, string password) that I am testing. It calls another method GetIdpAccessToken(string username, string password). I need to mock the response from GetIdpAccessToken so that I can create a few of the possible Idp responses to make sure the Login method code is doing what it is supposed to.

If there is no way to do this maybe it is showing me a design flaw in my code, maybe I need an IdpService?

Thanks for any help.

**Comment by markjsc** [comment]
Hi Jason (@jmichas),
This does sound like a design change is in order. I think you're on the right path with extracting the logic of the `GetIdpAccessToken` into a separate dependency/service with an interface (`IIdpAccessTokenService` or something). Then you can just provide the new interface as a constructor parameter of the class under test. And then you can easily mock the functionality of the new interface at test time.

A good example I've often encountered are File Access operations - reading or writing external files. These are not very unit-test-friendly - they're slow and take setup and tear-down to get the necessary files in place and to verify that they succeeded. (These are more appropriate for Integration tests.) I almost always end up with an `IFileAccess` interface where the implementation consists of just pass-through calls to the respective `System.IO.File` or `System.IO.FileInfo` classes. Then I can easily mock `IFileAccess` in my unit tests and I don't have to bother setting up anything in the file system at unit test time.

Does that answer your question?

Thanks for checking out AutoMockHelper and please let me know if I can make any improvements!
-Mark

Edited for clarity.

**Comment by jmichas** [comment]
Mark,

Thanks for getting back to me.
Yeah, I realized after trying a couple of different things that refactoring into a new service was the best way. But, that is what TDD is for and even though I am writing these tests after development it is good to help identify areas where the code is a little wonky.

Thanks for a great helper library, it has really saved me a ton of time over what I used to do manually creating mocks.

-Jason
